solarforecastarbiter/metrics/preprocessing.py

```diff
@@ -895,3 +963,143 @@ def _name_pfxobs(current_names, forecast, i=1):
         return _name_pfxobs(current_names, new_name, i + 1)
     else:
         return forecast_name
+
+
+def get_forecast_report_issue_times(
+    forecast: datamodel.Forecast,
+    start: pd.Timestamp,
+    end: pd.Timestamp
+) -> pd.DatetimeIndex:
+    """Returns all of the forecast issue times that contribute data
+    to a report. May include issue times that correspend with data
+    before and after the report to ensure report coverage.
+
+    Parameters
+    ----------
+    forecast: datamodel.Forecast
+        Forecast to find issue times for.
+    start: pd.Timestamp
+        Start of the report.
+    end: pd.Timestamp
+        End of the report.
+
+    Returns
+    -------
+    pandas.DatetimeIndex
+        Pandas DatetimeIndex representing all of the issue times.
+    """
+    # Get total forecast horizon to know how far back to look for issue times
+    total_forecast_horizon = forecast.lead_time_to_start + forecast.run_length
+
+    # Convert start to UTC so we can replace with the UTC-defined issue time
+    # to get the first issue time inside the lookback period.
+    lookback_start = start.tz_convert('UTC') - total_forecast_horizon
+    issue_start = lookback_start.replace(
+        hour=forecast.issue_time_of_day.hour
+    )
+
+    # If we undershot the lookback period when aligning with first issue
+    # time, look back another day so we can start from a day's first
+    # issue time without undershooting the forecast horizon.
+    if issue_start > lookback_start:
+        issue_start -= pd.Timedelta('1D')
+
+    # Get all possible issue times that contribute to the report
+    issue_times = pd.date_range(
+        issue_start,
+        end.tz_convert('UTC'),
+        freq=forecast.run_length
+    )
+    return issue_times
+
+
+def get_outage_periods(
+    forecast: datamodel.Forecast,
+    start: pd.Timestamp,
+    end: pd.Timestamp,
+    outages: Tuple[datamodel.TimePeriod, ...]
+) -> Tuple[datamodel.TimePeriod, ...]:
+    """Converts report outage periods to forecast data periods to
+    drop from analysis. The returned periods do not account for
+    interval label.
```

Comment:
makes me wonder if TimePeriod needs `interval_label` attribute to help with bookkeeping but perhaps we should move on for now. It would remove the need for the `interval_label` argument in `remove_outage_periods`

Reply:
The reason for waiting to apply an interval label until the `remove_outage_periods` is because the forecast interval label might be different than the observation's in a given pair.